Python/marshal.c

```diff
@@ -1954,28 +2026,25 @@ marshal_loads_impl(PyObject *module, Py_buffer *bytes, int lazy)
     if (lazy < 0)
         lazy = getenv("LAZY") && *getenv("LAZY");  // TODO: Rename?
     if (lazy) {
-        rf.ctx = PyMem_Malloc(sizeof(struct context));
+        rf.ctx = _PyHydrationContext_new(
+            bytes->obj, s, n, rf.refs);
+        Py_DECREF(rf.refs);
         if (rf.ctx == NULL) {
             PyErr_NoMemory();
             return NULL;
         }
-        Py_INCREF(bytes->obj);
-        rf.ctx->obj = bytes->obj;
-        rf.ctx->buf = s;
-        rf.ctx->len = n;
-        rf.ctx->code = NULL;
-        Py_INCREF(rf.refs);
-        rf.ctx->refs = rf.refs;
     }
     result = read_object(&rf);
-    Py_DECREF(rf.refs);
+    if (lazy) {
+        Py_DECREF(rf.ctx);
+    }
```

Comment:
This should just be XDECREF. Much better to check the actual state than to repeat the flag check.